Add global pooling specialization and also update MaxPooling on GPU

xiaomengy opened the pull request:
Summary: Add global pooling specialization and also update MaxPooling on GPU

Differential Revision: D13596340

houseroad commented on caffe2/operators/pool_op.cc:
```diff
@@ -721,6 +764,49 @@ CAFFE2_SPECIALIZED_AVERAGE_POOL_FUNCTOR_FORWARD_3D(float, StorageOrder::NCHW)
+    ConstEigenArrayMap<float> X_arr(X_ptr, C, HxW);
+    EigenVectorArrayMap<float> Y_arr(Y_ptr, C);
+    for (int j = 0; j < HxW; ++j) {
+      Y_arr = Y_arr.max(X_arr.col(j));
```
Do we really need Eigen support here? Seems we still use two loops to iterate over the maxtrix.

BIT-silence replied:
In NHWC order, here we are computing the elementwise max on a vector whose size is C. So eigen will help us implement SIMD.

BIT-silence replied:
Actually we cannot do RowwiseReduce here since the current impl of Rowwise Reduce in math is not efficient because of the problem in eigen. I will try to improve that later then we can use math function here.

houseroad commented on caffe2/operators/pool_op.cu:
```diff
@@ -253,7 +254,7 @@ __global__ void AveragePool3DForwardNHWCCUDAKernel(
   const int xw = yw * stride_w;
   const int p = max(xd - pad_p, 0);
-  const int a = min(xd - pad_p + kernel_h, X_D);
+  const int a = min(xd - pad_p + kernel_d, X_D);
```
We really need to improve our test coverage on ops.

BIT-silence replied:
Sure, I will update pooling_test later.

houseroad reviewed (commented):
I think you should also add some comments on how you split the data load between different threads. It's not very straightforward, and I forgot to mention that in the previous comments.

houseroad reviewed (approved):
Looks good!

xiaomengy committed:
Add global pooling specialization and also update MaxPooling on GPU (#15824)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15824

Add global pooling specialization and also update MaxPooling on GPU

Reviewed By: houseroad

Differential Revision: D13596340

fbshipit-source-id: b0044d6ed1107916c73f98cbca3ff32da21f485a
